keps/sig-node/20190226-pod-overhead.md

```diff
@@ -0,0 +1,310 @@
+- What changes (in invocations, configurations, API use, etc.) is an existing cluster required to make on upgrade in order to keep previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing cluster required to make on upgrade in order to make use of the enhancement?
+
+### Version Skew Strategy
+
+Set the overhead to the max of the two version until the rollout is complete.  This may be more problematic
+if a new version increases (rather than decreases) the required resources.
+
+## Implementation History
+
+Major milestones in the life cycle of a KEP should be tracked in `Implementation History`.
+Major milestones might include
+
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+
+## Drawbacks [optional]
+
+This KEP introduceds further complexity, and adds a field the PodSpec which users aren't expected to modify.
+
+## Alternatives [optional]
+
+In order to achieve proper handling of sandbox runtimes, the scheduler/resourceQuota handling needs to take
+into account the overheads associated with running a particular runtimeClass.
+
+### Introduce pod level resource requirements
+
+Rather than just introduce overhead, add support for general pod-level resource requirements. Pod level
+resource requirements are useful for shared resources (hugepages, memory when doing emptyDir volumes).
+
+Even if this were to be introduced, there is a benefit in keeping the overhead separate.
+ - post-pod creation handling of pod events: if runtimeClass definition is removed after a pod is created,
+  it will be very complicated to calculate which part of the pod resource requirements were associated with
+  the workloads versus the sandbox overhead.
+ - a kubernetes service provider can subsidize the charge-back model potentially and eat the cost of the
+ runtime choice, but charge the user for the cpu/memory consumed independent of runtime choice.
+
+
+### Leaving the PodSpec unchaged
+
+Instead of tracking the overhead associated with running a workload with a given runtimeClass in the PodSpec,
+the Kubelet (for pod cgroup creation), the scheduler (for honoring reqests overhead for the pod) and the resource
+quota handling (for optionally taking requests overhead of a workload into account) will need to be augmented
+to add a sandbox overhead when applicable.
+
+Pros:
+ * no changes to the pod spec
+ * no need for a mutating admission controller
+
+Cons:
+ * handling of the pod overhead is spread out across a few components
+ * Not user perceptible from a workload perspective.
+ * very complicated if the runtimeClass policy changes after workloads are running
```

Comment:
It's worth noting that we don't merge the runtime Handler into the pod spec. The RuntimeClass _is_ immutable, although it can still be deleted and recreated. We could add special handling to prevent the RuntimeClass from being deleted as long there are pods running with it.